### PR TITLE
feat(mbm): memory bandwidth management

### PR DESCRIPTION
**Pull request by h-w-chen**
#### What type of PR is this?
Feature: full mbm (memory bandwidth management) functionality

#### What this PR does / why we need it:
this PR includes 2 major parts:
1. memory bandwidth metrics collection: periodically collects NUMA/package level memory bandwidth/latency data, and saves to metric store for other component to use.
2. memory bandwidth adjustment: adjust mem bandwidth quotas of numa nodes in a physical package to ensure workloads' bandwidth not impacted by noisy neighbours (those consume too much bandwidth) based on specific  threshold value
 
##### Related Startup Args
By default this metric provisioner is disabled. To enable it, start up arg like below should be provided:
```
--metric-provisioners="...,mbw,..."
```
The interval (e.g. 1 second) to refresh these metrics is set as start up following arg (the default 5 seconds is not the typically desired value for these metrics):
```
--metric-provisioner-intervals mbw=1s
```

To enable mem bandwidth adjustment, and specify the adjustmant cycle interval 1 sec, the mem bandwidth threshold value in MB per second,
```
--enable-mbm --mbm-latency-threshold=14000 --mbm-control-interval=1s
```

#### Which issue(s) this PR fixes:
NA - new feature

#### Special notes for your reviewer:
this PR is arranged roughly in 3 blocks of commits, hopefully reviewers can pick the relevant bock to provide feedbacks:
1. ci job additions;
3. mbw lib refactorings (mbw metrics collection part)
4. metrics provisioner implementation
5. mbw lib mem bandwidth adjustment part
6. mbm adjustment control inside qrm plugin (the actual adjustment is via external manager)

### CODE REVIEW

File: pkg/mbw/monitor/monitor.go [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+)
+
+const (
+	MBM_MONITOR_INTERVAL = 1000
+
+	MAX_NUMA_DISTANCE = 32 // this is the inter-socket distance on AMD, intel inter-socket distance < 32
+	MIN_NUMA_DISTANCE = 11 // the distance to local numa is 10 in Linux, thus 11 is the minimum inter-numa distance
+)
+
+func newSysInfo(machineInfoConfig *global.MachineInfoConfiguration) (*SysInfo, error) {
```
Comment (luomingmeng) [inline pkg/mbw/monitor/monitor.go:39]: Can all the static sysInfo move to metaserver KatalystMachineInfo?

Comment (h-w-chen) [reply pkg/mbw/monitor/monitor.go:39]: yes this is a good suggestion. will refactor to make these into KatalystMachineInfo's cpu topology,  CCD topology respectively in next commit.

Comment (h-w-chen) [reply pkg/mbw/monitor/monitor.go]: code refactored and sysinfo structure is merged into katalyst machine info

File: pkg/mbw/monitor/controller.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+type MB_CONTROL_ACTION int
+
+const (
+	BW_LEN                = 11
+	MBA_COS_MAX           = 16 // AMD supports up to 16 cos per CCD, while Intel supports up to 16 cos per socket
+	L3QOS_BW_CONTROL_BASE = 0xc0000200
+
+	MEMORY_BANDWIDTH_THRESHOLD_PHYSICAL_NUMA = 140000 // fixed threshold 140GBps each physical numa node
```
Comment (smart2003) [inline pkg/mbw/monitor/controller.go:33]: I used this const for testing. Its value should be assigned from configuration and machine dependent. If this PR focuses on mbw monitoring only, this file can be removed.

File: pkg/mbw/monitor/pci.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	}
+
+	numaPerSocket, err := m.NUMAsPerSocket()
+	if err != nil {
+		general.Errorf("failed to get numa per socket - %v", err)
+		return err
+	}
+
+	// get all IOHC devs on PCI
+	devs := pci.ScanDevices(uint16(DRV_IS_PCI_VENDOR_ID_AMD), pciDevID)
```
Comment (smart2003) [inline pkg/mbw/monitor/pci.go:48]: throw an error if len(devs) == 0. It is generally caused by the absence of  libpci-dev。

File: pkg/mbw/monitor/umc.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+			// accumulate the UMC result for each package (i.e. physical Numa)
+			if (umc+1)%m.PMU.UMC.NumPerPackage == 0 {
+				pkgIdx := (skt*m.PMU.UMC.NumPerSocket + umc) / m.PMU.UMC.NumPerPackage
+				m.MemoryBandwidth.Packages[pkgIdx].RMB_Delta = utils.Delta(48, cas_rd, m.MemoryBandwidth.Packages[pkgIdx].RMB)
+				m.MemoryBandwidth.Packages[pkgIdx].WMB_Delta = utils.Delta(48, cas_wr, m.MemoryBandwidth.Packages[pkgIdx].WMB)
+				m.MemoryBandwidth.Packages[pkgIdx].Total = utils.PMUToMB(m.MemoryBandwidth.Packages[pkgIdx].WMB_Delta+m.MemoryBandwidth.Packages[pkgIdx].RMB_Delta, m.Interval)
+
+				m.MemoryBandwidth.Packages[pkgIdx].RMB = cas_rd
+				m.MemoryBandwidth.Packages[pkgIdx].WMB = cas_wr
+				// general.Infof("get the mb result for physical node %d: (%d, %d)", (skt*m.UMC.NumPerSocket+umc)/m.UMC.NumPerPackage, cas_rd, cas_wr)
```
Comment (smart2003) [inline pkg/mbw/monitor/umc.go:141]: either remove or uncomment it. same to line 131

**Review by smart2003 — COMMENTED** [review]
Generally LGTM.

File: pkg/mbw/monitor/pci.go (hunk 3 above)
Comment (h-w-chen) [reply pkg/mbw/monitor/pci.go:48]: will change code as suggested to treat such error properly

File: pkg/mbw/monitor/umc.go (hunk 4 above)
Comment (h-w-chen) [reply pkg/mbw/monitor/umc.go:141]: ok will remove all the commented log info lines

Comment (h-w-chen) [reply pkg/mbw/monitor/umc.go]: ok will remove commented out log lines

File: pkg/util/machine/die.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machine
+
+// DieTopology keeps the relationship of dies(CCDs), numa, package, and socket
```
Comment (csfldf) [inline pkg/util/machine/die.go:19]: we will schedule a meeting to make agreement on that if it's better to make the topology more generic cc @luomingmeng

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+			return nil, err
+		}
+		groups = append(groups, groupMB)
+	}
+	return groups, nil
+}
+
+func (c Controller) processPackage(packageID int, nodes []int) {
+	// get the metrics
+	currMetric, err := c.metricReader.GetPackageMetric(packageID, "rw-mbps")
```
Comment (csfldf) [inline pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go:132]: define "rw-mbps" as consts in pkg/consts/metric.go

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+func (c Controller) run() {
+	for p, nodes := range c.packageMap {
+		c.processPackage(p, nodes)
+	}
+}
+
+// getActiveNodeSets divides the nodes into workload-associated sets
+// only nodes having active workload assigned shall be considered
+func (c Controller) getActiveNodeSets(nodes []int) map[string]sets.Int {
```
Comment (csfldf) [inline pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go:78]: filter dedicated_cores with numa_binding pods

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+		groups = append(groups, groupMB)
+	}
+	return groups, nil
+}
+
+func (c Controller) processPackage(packageID int, nodes []int) {
+	// get the metrics
+	currMetric, err := c.metricReader.GetPackageMetric(packageID, "rw-mbps")
+	if err != nil {
+		c.metricEmitter.StoreInt64("package metric reading error", 1, metrics.MetricTypeNameCount)
```
Comment (csfldf) [inline pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go:134]: replace "package metric reading error" with a metric key defined in qrm-plugins/util/consts.go

all similar places

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+			for node, share := range group {
+				// ignore if the deduction step is too small
+				if share >= c.minDeductionStep {
+					c.numaThrottled.Insert(node)
+					if err := c.mbAdjust.AdjustNumaMB(node,
+						uint64(activeGroupMBs.NodeMBAverage()),
+						uint64(share),
+						mbm.MEMORY_BANDWIDTH_CONTROL_REDUCE,
+					); err == nil {
+						c.numaThrottled.Insert(node)
```
Comment (csfldf) [inline pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go:168]: c.numaThrottled.Insert(node) has beed called at line 162? Is it duplicated?

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/strategy/relu_allocator.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+limitations under the License.
+*/
+
+package strategy
+
+// reluAllocator is the reLU style allocator to decide the shares among the participating numa nodes
+type reluAllocator struct{}
+
+func (r reluAllocator) AllocateDeductions(activeGroups GroupMBs, packageMB, mbThreshold float64) []map[int]float64 {
+	excessiveByPackage := packageMB - mbThreshold
```
Comment (csfldf) [inline pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/strategy/relu_allocator.go:23]: should we identify the case that excessiveByPackage is negtive and return a error ?

File: pkg/metaserver/external/manager_linux.go [hunk 11]
```diff
@@ -92,3 +95,10 @@ func (m *externalManagerImpl) setComponentImplementation(setter func()) {
 
 	setter()
 }
+
+func (m *externalManagerImpl) AdjustNumaMB(node int, avgMB, quota uint64, action mbm.MB_CONTROL_ACTION) error {
+	if mbw.MBAdjuster == nil {
+		return errors.New("low level mbw not enabled")
+	}
+	return mbw.MBAdjuster.AdjustNumaMB(node, avgMB, quota, action)
```
Comment (csfldf) [inline pkg/metaserver/external/manager_linux.go:103]: why not implement MBAdjuster as a property of externalManagerImpl, just like NetworkManager or CgroupIDManager

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go (hunk 6 above)
Comment (h-w-chen) [reply pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go:132]: sure; and will change other string literals in similar way.

File: pkg/metaserver/external/manager_linux.go (hunk 11 above)
Comment (h-w-chen) [reply pkg/metaserver/external/manager_linux.go:103]: good advice. will update code as suggested.

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/strategy/relu_allocator.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// reluCalcShares distributes the shares of total to each slot based on the current in use amount, e.g.
+// [{0:100}, {1:150}, {2:80}] : current bandwidths of 3 numa nodes, the target deduction 40,
+// [{0:  2}, {1: 38], {2: 0}] : the optimal share distributions of deduction 40,
+// [{0: 98}, {1:112}, {2:80}] : the next in-use quotas,
+// as the bar being noisy neighbor is ((100+150+80)-40) / 3 = 97; a more complicated example is
+// [{4:110}, {3:160, 5:20}]: current bandwidths of 2 groups (including a 2-node group), total to decrease 20,
+// [{4: 20}, {3:  0, 5: 0}], as the set [3,5] consumes no more than they are entitled to combined.
+func (r reluAllocator) reluCalcShares(targetDeduction float64, useGroups GroupMBs) []map[int]float64 {
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/strategy/relu_allocator.go:37]: it is possible to do so. I'll add a code comment for future consideration.

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/strategy/relu_allocator.go (hunk 10 above)
Comment (h-w-chen) [reply pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/strategy/relu_allocator.go:23]: not really needed here as this method is called only by mbm controller and the controller (caller) should make sure the condition is met. The same applies to AllocateIncreases method too. I'll add comments to interface signature to make this assumption sort of clear.

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go (hunk 9 above)
Comment (h-w-chen) [reply pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go:168]: yes it is. line 168 should be kept as it has error checked. will update code.

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go (hunk 8 above)
Comment (h-w-chen) [reply pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go:134]: ok will follow the suite of other places, with the key indicating failures of package/numa metrics readings, and tag name being the metric name to get

File: pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go (hunk 7 above)
Comment (h-w-chen) [reply pkg/agent/qrm-plugins/cpu/dynamicpolicy/mbm/control.go:78]: for now only socket type pods are assumed on the machine - which is a limitation. Will add support of mixed pod types in the subsequent change AFTER the current code be validated. I will add comments to code to make this assumption and plan clear. Let me know if code comment not good enough in this case.

File: pkg/util/machine/extension.go [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+				return i
+			}
+		}
+	}
+
+	general.Errorf("failed to find the corresponding package for numa %d", numa)
+	return -1
+}
+
+func (s *KatalystMachineInfo) GetPackageMap(numasPerPackage int) map[int][]int {
```
[the remark this answers is not in the document]
Comment (h-w-chen) [reply pkg/util/machine/extension.go:122]: sure will use SiblingNumaInfo to derive such package->numa mapping

**Comment by h-w-chen** [comment]
will use resctrl to monitor mb. development plan changed.

**Pull request closed without merging**